spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -88,82 +93,247 @@ public void setTaskExecutor(TaskExecutor taskExecutor) {
+	 */
+	@Override
+	public final void setExclusive(boolean exclusive) {
+		Assert.isTrue(!exclusive || (this.consumersPerQueue == 1),
+				"When the consumer is exclusive, the consumers per queue must be 1");
+		super.setExclusive(exclusive);
+	}
+
+	/**
+	 * Set the interval before retrying when starting the container; default 5000.
+	 * @param recoveryInterval the interval.
 	 */
-	public void setDefaultRequeueRejected(boolean requeueRejected) {
-		this.defaultRequeueRejected = requeueRejected;
+	public void setRecoveryInterval(long recoveryInterval) {
+		this.recoveryInterval = recoveryInterval;
+	}
+
+	@Override
+	public void setQueueNames(String... queueName) {
+		Assert.state(!isRunning(), "Cannot set queue names while running, use add/remove");
+		super.setQueueNames(queueName);
+	}
+
+	@Override
+	public void setQueues(Queue... queues) {
+		Assert.state(!isRunning(), "Cannot set queue names while running, use add/remove");
+		super.setQueues(queues);
+	}
+
+	@Override
+	public void addQueueNames(String... queueNames) {
+		try {
+			addQueues(Arrays.stream(queueNames));
+		}
+		catch (AmqpIOException e) {
+			throw new AmqpIOException("Failed to add " + queueNames, e.getCause());
+		}
+		super.addQueueNames(queueNames);
 	}
 
 	@Override
+	public void addQueues(Queue... queues) {
+		try {
+			addQueues(Arrays.stream(queues)
+					.map(Queue::getName));
+		}
+		catch (AmqpIOException e) {
+			throw new AmqpIOException("Failed to add " + queues, e.getCause());
+		}
+		super.addQueues(queues);
+	}
+
+	private void addQueues(Stream<String> queueNameStream) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				Set<String> current = getQueueNamesAsSet();
+				List<String> added = new ArrayList<>();
+				queueNameStream.forEach(queue -> {
+					if (current.contains(queue)) {
+						this.logger.warn("Queue " + queue + " is already configured for this container: "
+								+ this + ", ignoring add");
+					}
+					else {
+						try {
+							consumeFromQueue(queue);
+							added.add(queue);
+						}
+						catch (IOException e) {
+							logger.error("Failed to add queue: " + queue + " undoing adds (if any)");
+							removeQueues(added.stream());
+							throw new AmqpIOException(e);
+						}
+					}
+				});
+			}
+		}
+	}
+
+	@Override
+	public boolean removeQueueNames(String... queueNames) {
+		removeQueues(Arrays.stream(queueNames));
+		return super.removeQueueNames(queueNames);
+	}
+
+	@Override
+	public boolean removeQueues(Queue... queues) {
+		removeQueues(Arrays.asList(queues)
+				.stream()
+				.map(Queue::getName));
+		return super.removeQueues(queues);
+	}
+
+	private void removeQueues(Stream<String> queueNames) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				queueNames.forEach(queue -> {
+					List<SimpleConsumer> consumersForQueue = this.consumersByQueue.remove(queue);
+					if (consumersForQueue != null) {
+						consumersForQueue.stream().forEach(consumer -> {
```

Comment:
IDEA says me that `.stream()` is redundant here and `.forEach()` is a part of `Iterable` API.